rfcs/text/0000-data_stream-fields.md

```diff
@@ -0,0 +1,158 @@
+# 0000: Data stream fields
+<!-- Leave this ID at 0000. The ECS team will assign a unique, contiguous RFC number upon merging the initial stage of this RFC. -->
+
+- Stage: **1 (proposal)** <!-- Update to reflect target stage. See https://elastic.github.io/ecs/stages.html -->
+- Date: **TBD** <!-- The ECS team sets this date at merge time. This is the date of the latest stage advancement. -->
+
+When introducing the new indexing strategy for Elastic Agent which uses data streams, we found that adding a few "constant_keyword" fields corresponding to the central components in the new indexing strategy would be advantageous. 
+
+
+<!--
+As you work on your RFC, use the "Stage N" comments to guide you in what you should focus on, for the stage you're targeting.
+Feel free to remove these comments as you go along.
+-->
+
+<!--
+Stage 0: Provide a high level summary of the premise of these changes. Briefly describe the nature, purpose, and impact of the changes. ~2-5 sentences.
+-->
+
+## Fields
+
+<!--
+Stage 1: Describe at a high level how this change affects fields. Which fieldsets will be impacted? How many fields overall? Are we primarily adding fields, removing fields, or changing existing fields? The goal here is to understand the fundamental technical implications and likely extent of these changes. ~2-5 sentences.
+-->
+
+This RFC proposes to introduce a new fieldset called "data_stream". The fieldset consists of the following fields:
+Field     | Mapping type | Description
+----------|--------------|--------------
+data_stream.type | constant_keyword | An overarching type for the data stream. Currently allowed values include "logs", "metrics". We expect to also add "traces" and "synthetics" in the near future
+data_stream.dataset | constant_keyword | The field can contain anything that makes sense to signify the source of the data. Examples include `nginx.access`, `prometheus`, `endpoint` etc. For data streams that otherwise fit, but that do not have dataset set we use the value "generic" for the dataset value. `event.dataset` should have the same value as `data_stream.dataset`. 
+data_stream.namespace | constant_keyword | A user defined namespace. Namespaces are useful to allow grouping of data. Many of our customers already organize their indices this way, and now we are providing this best practice as a default. Many people will use `default` as the value.
```

Comment:
I had always assumed the namespace to be for separation between environments such as production and staging.

Although I see an example below of "web-frontend", which opens up very interesting possibilities.

I'm looking more for a brainstorm than requesting a change here. But would it be fair for users to use namespace for:

- separation of environments (as I initially assumed): `logs-nginx.access-prod`, `logs-nginx.access-staging`
- separate different applications/systems using the same technology: `logs-nginx.access-ecom`,  `logs-nginx.access-website`, `logs-nginx.access-intranet`
- separate different tiers of an application: `logs-nginx.access-web-frontend`,  `logs-nginx.access-web-backend`
- separate different parts of a complex system using the same technology: `logs-nginx.access-ecom-api`, `logs-nginx.access-ecom-admin`, `logs-nginx.access-ecom-sites`
- combine some of the approaches above as needed

Is that the intent? Have other potential uses been floated?

Reply:
indeed, the intent is for namespace to be the escape hatch that lets user groups their data in any way they see fit - in addition to the very basic grouping we achieve through `data_stream.type` and `data_stream.dataset`.

The idea is that we provide a framework for naming your data streams without taking away much flexibility with regard to grouping your data.